ddtrace/internal/forksafe.py

```diff
@@ -0,0 +1,106 @@
+import inspect
+import functools
+import os
+import threading
+
+
+__all__ = [
+    "call_nocheck",
+    "ddtrace_after_in_child",
+    "register",
+]
+
+
+BUILTIN_FORK_HOOKS = hasattr(os, "register_at_fork")
+
+if BUILTIN_FORK_HOOKS:
+    registry = []
+
+    def ddtrace_after_in_child():
+        global registry
+
+        for fn, hook in registry:
+            hook()
+            fn._fork_occurred = True
+
+    def register(after_in_child):
+        # TODO: this function won't work if not used as a decorator.
+        # eg: register(hook) won't result in hook being called.
+        def wrapper(func):
+            args = inspect.getfullargspec(func)[0]
+            include_arg = "is_in_child_after_fork" in args
```

Comment:
I understand why you did that, but I don't think it's a good stance. This makes things complicated (looking into signature is really not a great play) and might bite us with race condition, e.g. you have 2 threads in the fork, what happens? Which one plays with `_fork_occured` first?
I don't think that this is necessary. At worst you just need to re-architect a little bit the `Tracer` code to move the code in the hook.

We shouldn't build too much machinery around that.

Reply:
Hmm. I agree that the implementation is unorthodox and it doesn't quite sit well with me. As for the race condition we could synchronize it with a lock, whichever function gets called first would get to handle `is_in_child_after_fork`. However, I think you might be right that we can avoid this weirdness.